content/en/blog/_posts/2024-08-15-volume-attributes-class.md

```diff
@@ -0,0 +1,99 @@
+Kubernetes has no API which allows changing them.
+
+We are pleased to announce that the [VolumeAttributesClass
+KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/3751-volume-attributes-class/README.md),
+alpha since Kubernetes 1.29, will be beta in 1.31. This provides a generic,
+Kubernetes-native API for modifying volume parameters like provisioned IO.
+
+Like all new volume features in Kubernetes, this API is implemented via the [container
+storage interface (CSI)](https://kubernetes-csi.github.io/docs/). In addition to the
+VolumeAttributesClass feature gate, your provisioner-specific CSI driver must support the
+new ModifyVolume API which is the CSI side of this feature.
+
+See the [full
+documentation](https://kubernetes.io/docs/concepts/storage/volume-attributes-classes/)
+for all details. Here we show the common workflow.
+
+### Dynamically modifying volume attributes.
+
+A `VolumeAttributesClass` is a cluster-scoped resource that specifies provisioner-specific
+attributes. These are created by the cluster administrator in the same way as storage
+classes. For example, a series of gold, silver and bronze volume attribute classes can be
+created for volumes with greater or lessor amounts of provisioned IO.
+
+```yaml
+apiVersion: storage.k8s.io/v1alpha1
+kind: VolumeAttributesClass
+metadata:
+  name: silver
+driverName: pd.csi.storage.gke.io
+parameters:
+  provisioned-iops: "500"
+  provisioned-throughput: "50MiB/s"
+---
+apiVersion: storage.k8s.io/v1alpha1
+kind: VolumeAttributesClass
+metadata:
+  name: silver
+driverName: pd.csi.storage.gke.io
+parameters:
+  provisioned-iops: "10000"
+  provisioned-throughput: "500MiB/s"
+```
+  
+An attribute class is added to a PVC in much the same way as a storage class.
+
+```yaml
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: test-pv-claim
+spec:
+  storageClassName: csi-sc-example
+  volumeAttributesClassName: silver
+  accessModes:
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: 64Gi
+```
+
+Unlike a storage class, the volume attributes class can be changed:
+
+```
+kubectl patch pvc test-pv-claim -p '{"spec": "volumeAttributesClassName": "gold"}'
+```
+
+Kubernetes will work with the provisioner CSI driver to update the attributes of the
+volume. The status of the PVC will track the current and desired attributes class. The PV
+resource has also been updated with a volume attributes class which will be set to the
+currently active attributes of the PV.
+
+### Limitations with the beta
+
+As a beta feature, there are still some features which are planned for GA but not yet
+present. The largest is quota support, see the
+[KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/3751-volume-attributes-class/README.md)
+and discussion in
+[sig-storage](https://github.com/kubernetes/community/tree/master/sig-storage) for details.
+
+Support for this feature in the `pd.csi.storage.gke.io` driver is in process of
```

Comment:
It's probably better to link to the csi drivers documentation to see the list of drivers that currently support the feature.: https://kubernetes-csi.github.io/docs/drivers.html

Reply:
Done. I wanted to point out the pd csi driver particularly, since we (unforunately & ironically) use that in our examples.

Reply:
(and, TIL we should update that page when we finish the VAC support :-)

Reply:
Maybe we should just replace driver name with some generic "my-csi-driver" in the examples then?

Reply:
Yeah, good idea.